internal/pkg/agent/application/coordinator/state/state.go

```diff
@@ -0,0 +1,365 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package state
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	agentclient "github.com/elastic/elastic-agent/pkg/control/v2/client"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/client"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent/pkg/component/runtime"
+)
+
+// State provides the current state of the coordinator along with all the current states of components and units.
+type State struct {
+	State        agentclient.State                 `yaml:"state"`
+	Message      string                            `yaml:"message"`
+	FleetState   agentclient.State                 `yaml:"fleet_state"`
+	FleetMessage string                            `yaml:"fleet_message"`
+	Components   []runtime.ComponentComponentState `yaml:"components"`
+	LogLevel     logp.Level                        `yaml:"log_level"`
+}
+
+// StateFetcher provides an interface to fetch the current state of the coordinator.
+type StateFetcher interface {
+	// State returns the current state of the coordinator.
+	State() State
+}
+
+type CoordinatorState struct {
+	mx            sync.RWMutex
+	state         agentclient.State
+	message       string
+	fleetState    agentclient.State
+	fleetMessage  string
+	logLevel      logp.Level
+	overrideState *coordinatorOverrideState
+
+	compStatesMx sync.RWMutex
+	compStates   []runtime.ComponentComponentState
+
+	mgrMx         sync.RWMutex
+	runtimeMgrErr error
+	configMgrErr  error
+	actionsErr    error
+	varsMgrErr    error
+
+	subMx     sync.RWMutex
+	subscribe []*StateSubscription
+}
+
+type coordinatorOverrideState struct {
+	state   agentclient.State
+	message string
+}
+
+type stateSetter struct {
+	state        *agentclient.State
+	message      *string
+	fleetState   *agentclient.State
+	fleetMessage *string
+	logLevel     *logp.Level
+}
+
+type stateSetterOpt func(ss *stateSetter)
+
+// NewCoordinatorState creates the coordinator state manager.
+func NewCoordinatorState(state agentclient.State, msg string, fleetState agentclient.State, fleetMsg string, logLevel logp.Level) *CoordinatorState {
+	return &CoordinatorState{
+		state:        state,
+		message:      msg,
+		fleetState:   fleetState,
+		fleetMessage: fleetMsg,
+		logLevel:     logLevel,
+	}
+}
+
+// UpdateState updates the state triggering a change notification to subscribers.
+func (cs *CoordinatorState) UpdateState(setters ...stateSetterOpt) {
+	var setter stateSetter
+	for _, ss := range setters {
+		ss(&setter)
+	}
+
+	cs.mx.Lock()
+	if setter.state != nil {
+		cs.state = *setter.state
+	}
+	if setter.message != nil {
+		cs.message = *setter.message
+	}
+	if setter.fleetState != nil {
+		cs.fleetState = *setter.fleetState
+	}
+	if setter.fleetMessage != nil {
+		cs.fleetMessage = *setter.fleetMessage
+	}
+	if setter.logLevel != nil {
+		cs.logLevel = *setter.logLevel
+	}
+	cs.mx.Unlock()
```

Comment:
I get that the previous lines are simple assignments but can we put the `mutex.Unlock()`  in a `defer` just after acquiring the lock ?
(same for all the other occurrences)

Reply:
No it cannot be done with defer because `cs.mx.Unlock()` must be called before `cs.changed()` can be called.

Reply:
What about something like:

```suggestion
        func() {
		cs.mx.Lock()
		defer cs.mx.Unlock()
		if setter.state != nil {
			cs.state = *setter.state
		}
		if setter.message != nil {
			cs.message = *setter.message
		}
		if setter.fleetState != nil {
			cs.fleetState = *setter.fleetState
		}
		if setter.fleetMessage != nil {
			cs.fleetMessage = *setter.fleetMessage
		}
		if setter.logLevel != nil {
			cs.logLevel = *setter.logLevel
		}
	}()
```

Reply:
I think that is even harder to read.